src/udp/socket.rs

```diff
@@ -518,264 +535,234 @@ quick_error! {
-                            future::err(HolePunchError::UnexpectedMessage).into_boxed()
-                        }
-                    }
-                }
+            match secure_deserialise(&bytes, &self.their_pk, &self.our_sk) {
+                Ok(msg) => return Ok(Async::Ready(msg)),
+                Err(e) => {
+                    warn!("unreceived unrecognisable data on hole punching socket: {}", e);
+                },
             }
-        })
-        .into_boxed()
+        }
+    }
 }
 
-// send an ack to the other peer, then complete hole-punching from there.
-fn send_ack(
-    mut ctx: HolePunching,
-    socket: WithAddress,
-) -> BoxFuture<(WithAddress, bool), HolePunchError> {
-    trace!("sending ack to {}", socket.remote_addr());
+impl Future for HolePunching {
+    type Item = (WithAddress, bool);
+    type Error = HolePunchError;
 
-    ctx.send_msg(&HolePunchMsg::Ack, socket)
-        .and_then(move |socket| {
-            let try = || {
-                let socket = ctx.inc_ttl_and_syn_acks(socket)?;
-                Ok(recv_from_ack(ctx, socket))
+    fn poll(&mut self) -> Result<Async<(WithAddress, bool)>, HolePunchError> {
+        loop {
+            match self.flush()? {
+                Async::NotReady => return Ok(Async::NotReady),
+                Async::Ready(()) => (),
             };
-            future::result(try().map_err(HolePunchError::SetTtl)).flatten()
-        })
-        .into_boxed()
-}
 
-// we have just sent an ack to the peer, listen for incoming packets until it's time to send
-// another one.
-fn recv_from_ack(
-    ctx: HolePunching,
-    socket: WithAddress,
-) -> BoxFuture<(WithAddress, bool), HolePunchError> {
-    socket
-        .with_timeout_at(ctx.msg_timeout(), &ctx.handle)
-        .into_future()
-        .map_err(|(e, _)| HolePunchError::ReadMessage(e))
-        .and_then(move |(msg_opt, socket_timeout)| {
-            let socket = socket_timeout.into_inner();
-            match msg_opt {
-                None => send_ack(ctx, socket),
-                Some(recv_msg) => {
-                    match ctx.decrypt_msg(&recv_msg) {
-                        Err(e) => {
-                            warn!("error deserializing packet from peer: {:?}", e);
-                            recv_from_ack(ctx, socket)
+            if let Async::Ready(()) = self.timeout.poll().void_unwrap() {
+                self.timeout.reset(Instant::now() + Duration::from_millis(HOLE_PUNCH_MSG_PERIOD_MS));
+                let msg = match self.phase {
+                    HolePunchingPhase::Syn {
+                        ref mut time_of_last_ttl_increment,
+                        ttl_increment_duration,
+                    } => {
+                        let now = Instant::now();
+                        while now - *time_of_last_ttl_increment > ttl_increment_duration {
+                            let ttl = unwrap!(self.socket.as_mut()).ttl().map_err(HolePunchError::GetTtl)?;
```

Comment:
If you implemented `socket.inc_ttl()` that would look much more neat :)

Reply:
I had that originally but I had to inline it to appease the borrow-checker. I can't wait for non-lexical lifetimes :p